apturicovid/api/StatsClient.swift

```diff
@@ -4,71 +4,93 @@ import RxSwift
 class StatsClient: RestClient {
     
     /// Specifies how long stats data is valid from .updatedAt
-    private static let statTtlInterval: TimeInterval = 2 * 24 * 60 * 60 // 2 days
+    static let statTtlInterval: TimeInterval = 2 * 24 * 60 * 60 // 2 days
     
     private static let fetchThrottleWifi: TimeInterval = 5 * 60 // 5 min
     private static let fetchThrottleCellular: TimeInterval = 10 * 60 // 10 min
     
     static let shared = StatsClient()
     private override init() { }
     
-    func getStats(forceFromApi: Bool = false, ignoreOutdated: Bool = false) -> Observable<Stats> {
+    enum FetchSource {
+        case Api, Auto, Local
+    }
+    
+    func getStats(from source: FetchSource = .Auto, ignoreOutdated: Bool = false) -> Observable<Stats> {
         
-        return Observable<Stats>.create { [weak self] observer -> Disposable in
-            
-            let disposable = Disposables.create()
-            guard let `self` = self else { return disposable }
-            
-            // Return stored data
-            if !forceFromApi {
-                if let stats = LocalStore.shared.stats {
-                    
-                    // Skip outdated if required data
-                    if ignoreOutdated {
-                        let outdated = stats.updatedAt.distance(to: Date()) > StatsClient.statTtlInterval
-                        if !outdated { observer.onNext(stats) }
-                    } else {
-                        observer.onNext(stats)
-                    }
-                }
+        let statsObservable: Observable<Stats>
+        
+        switch source {
+        case .Api:
+            statsObservable = getStatsFromApi(forced: true)
+        case .Auto:
+            statsObservable = Observable.concat([getStatsFromLocalStorage(),getStatsFromApi(forced: false)])
+        case .Local:
+            statsObservable = getStatsFromLocalStorage()
+        }
+        
+        return statsObservable.filter{ stats in
+            ignoreOutdated ? !stats.isOutdated : true
+        }
+    }
+    
+    private func shouldFetchFromApi() -> Bool {
+        
+        guard let stats = LocalStore.shared.stats else {
```

Comment:
Doesn't this make always fetch from API once the first fetch has completed?

Reply:
No, guard statement guards from missing stats. If stats are available, fetch will be preferred if stats are outdated or throttle interval reached.